lua/lspconfig/server_configurations/drools_lsp.lua

```diff
@@ -1,35 +1,60 @@
 local util = require 'lspconfig.util'
 
-local get = function(config, section, key)
-  return vim.tbl_get(config.settings, section, key) and config.settings[section][key] or nil
+local function get(config, segments)
+  if vim.tbl_get(config, unpack(segments)) then
+    local cfg = config
+    for _, seg in ipairs(segments) do
+      cfg = cfg[seg]
+    end
+    return cfg
+  end
+  return nil
 end
 
-local get_java_bin = function()
-  local java_bin = vim.env.JAVA_HOME and util.path.join(vim.env.JAVA_HOME, 'bin', 'java') or 'java'
-  if vim.fn.has 'win32' == 1 then
-    java_bin = java_bin .. '.exe'
+local function get_java_bin(config)
+  local java_bin = get(config, { 'drools', 'java', 'bin' })
+  if java_bin == nil then
```

Comment:
```suggestion
  if vim.tbl_isempty(java_bin) then
```

Reply:
@glebnir, java_in is not a table, it's a string value that I'm looking for

Reply:
the get function your write return is `table|nil`

Reply:
The get function iterates over all the unpacked arguments sent to it, traversing down each node of the config until it gets to the last node that is being requested of it. For java_bin, it's not going to be a table value. It's going to be a string value.

Reply:
like i said you returned it with two type maybe one type is enough

Reply:
I'm sorry I do not understand what you're trying to say. We might be hitting a language barrier here.

The get function first checks if there is a value defined in the config that matches the segments "path" (using tbl_get and unpack). If not, it returns nil. If it does, then it can safely traverse down to get the value, which could be a string or a table. So the return of the get function is either nil (user didn't configure that node), a string (like java_bin), or a table (like cmd).

I don't know how to explain it better than that. If you don't like the implementation of the get function, perhaps you could suggest a complete get function that would replace it? Because otherwise I'm having trouble understanding what you are suggesting that I should do.

Very respectfully, I'm also confused as to why you consider the function is somehow "wrong" the way it is written? It works well, so these suggestions feel more like personal style preferences vs. fixing any actual bugs. I do very much appreciate the time you have spent on this, though. Thank you!

Reply:
>  I'm also confused as to why you consider the function is somehow "wrong" the way it is written?

 ```lua
local function get(config, segments)
  if vim.tbl_get(config, unpack(segments)) then
   -- cfg type is table
    local cfg = config
    for _, seg in ipairs(segments) do
      cfg = cfg[seg]
    end
    return cfg
  end
 --- return nil in there
  return nil
end
```
this function signature help return type will be `table or nil`. I say can we just return one type `table` my implment is one type

Reply:
The return type will be a table, a string, or nil. The value of `cfg` _starts out_ as a table, but the last seg in the segments loop could be the key to a table OR a string.

So, when looping over `{ 'drools', 'java', 'opts' }`, yes, the return type is a table. However, when looping over `{ 'drools', 'java', 'bin' }`, the return type is a string. So "returning just on type `table`" will not work.

Reply:
`{ 'drools', 'java', 'opts' }` in `vim.tbl_get` mean `config['drools']['java']['opts']` what does this code mean ?

```lua
   local cfg = config
   for _, seg in ipairs(segments) do
     cfg = cfg[seg]
   end
```
the cfg actually equals cofig['opts']  ? you set value to it in a for iterator the value will be the. last index of segments.

Reply:
If you take this invocation: `get(config, { 'drools', 'java', 'opts' })`, this is what happens:

`local cfg = config`
-- ^^ the value of `cfg` _starts out as_ the root of the config (`config`)

`for` loop, where `seg` currently equals `'drools'`
-- ^^ 1st iteration in loop,  `cfg` is a table

`for` loop, where `seg` currently equals `'java'`
-- ^^ 2nd iteration in loop,  `cfg` is a table

`for` loop, where `seg` currently equals `'opts'`
-- ^^ 3rd iteration in loop,  `cfg` _ends up as_ a table, _and we return that table_.

Now if you take this invocation: `get(config, { 'drools', 'java', 'bin' })`, this is what happens:

`local cfg = config`
-- ^^ the value of `cfg` _starts out as_ the root of the config (`config`)

`for` loop, where `seg` currently equals `'drools'`
-- ^^ 1st iteration in loop,  `cfg` is a table

`for` loop, where `seg` currently equals `'java'`
-- ^^ 2nd iteration in loop,  `cfg` is a table

`for` loop, where `seg` currently equals `'bin'`
-- ^^ 3rd iteration in loop,  `cfg` _ends up as_ a string, _and we return that string_.

I have tested this, and it is the behavior I expect and want.

Reply:
The value of cfg keeps getting reassigned to a deeper level, and returns when it is at the level that is being requested, whether the value at the level is a table or a string.

Reply:
The `get` function is so that I don't have to write this ugly code for every configuration value I'm looking for:
`local java_bin = vim.tbl_get(config, 'drools', 'java', 'bin') and config['drool']['java']['bin'] or nil`

Reply:
.oh you change the cfg table pointer. my mistake.  then we can simply use `vim.tbl_get` it will return this value . in doc exmaple it return true . so I think it just check . but I look at the source code .it will return the value .

`:lua print(vim.tbl_get({ a = { b = { c= 10}}}, 'a', 'b','c'))`   it will print 10

Reply:
It looks like using `vim.tbl_get` is also safe if `b` is nil, and returns nil instead of an error, which is awesome. So... If I get rid of the `get` function, and use `vim.tbl_get`, will that be satisfactory?

Reply:
yea. if key not exist it only return nil not trigger debug stack so just check nil is enought.